src/optimizer/rule_query_rewrite.cpp

```diff
@@ -0,0 +1,191 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// transitive_predicates.cpp
+//
+// Identification: src/optimizer/rule_query_rewrite.cpp
+//
+// Copyright (c) 2015-18, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include <memory>
+
+#include "catalog/column_catalog.h"
+#include "catalog/index_catalog.h"
+#include "catalog/table_catalog.h"
+#include "optimizer/rule_query_rewrite.h"
+#include "optimizer/operators.h"
+#include "optimizer/optimizer_metadata.h"
+#include "optimizer/properties.h"
+#include "optimizer/rule_impls.h"
+#include "optimizer/util.h"
+
+namespace peloton {
+namespace optimizer {
+
+//===--------------------------------------------------------------------===//
+// Transitive predicates rules
+//===--------------------------------------------------------------------===//
+
+///////////////////////////////////////////////////////////////////////////////
+/// LogicalGet
+TransitivePredicatesLogicalGet::TransitivePredicatesLogicalGet() {
+  type_ = RuleType::TRANSITIVE_PREDICATES_LOGICAL_GET;
+
+  match_pattern = std::make_shared<Pattern>(OpType::Get);
+}
+
+bool TransitivePredicatesLogicalGet::Check(std::shared_ptr<OperatorExpression> input,
+					   OptimizeContext *context) const {
+  (void)input;
+  (void)context;
+  return true;
+}
+
+void TransitivePredicatesLogicalGet::Transform(
+    std::shared_ptr<OperatorExpression> input,
+    std::vector<std::shared_ptr<OperatorExpression>> &transformed,
+    OptimizeContext *context) const {
+  LOG_TRACE("TransitivePredicatesLogicalGet::Transform");
+
+  auto get = input->Op().As<LogicalGet>();
+  auto &get_predicates = get->predicates;
+
+  util::FillTransitiveTable(get_predicates, context->transitive_table);
+  auto new_predicates = util::GenerateTransitivePredicates(get_predicates, context->transitive_table);
+
+  if (new_predicates.empty())
+    return;
+
+  std::vector<AnnotatedExpression> total_predicates;
+
+  for (auto &predicate : get_predicates) {
+    total_predicates.emplace_back(predicate);
+  }
+
+  for (auto &predicate : new_predicates) {
+    total_predicates.emplace_back(predicate);
+  }
+  
+  std::shared_ptr<OperatorExpression> output =
+      std::make_shared<OperatorExpression>(
+          LogicalGet::make(get->get_id, total_predicates,
+  			   get->table, get->table_alias, get->is_for_update));
+  
+  transformed.push_back(output);
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/// LogicalFilter
+TransitivePredicatesLogicalFilter::TransitivePredicatesLogicalFilter() {
+  type_ = RuleType::TRANSITIVE_PREDICATES_LOGICAL_FILTER;
+
+  std::shared_ptr<Pattern> child(std::make_shared<Pattern>(OpType::InnerJoin));
```

Comment:
Why is the child for this a join and not a leaf? Is there a reason we should only apply this to filters with joins as children?